### PR TITLE
feat(cluster): Migration cancellation support

### PR DESCRIPTION
**Pull request by chakaz**

### CODE REVIEW

File: src/server/cluster/incoming_slot_migration.cc [hunk 1]
```diff
@@ -85,15 +101,23 @@ IncomingSlotMigration::IncomingSlotMigration(string source_id, Service* se, Slot
...
 
 void IncomingSlotMigration::Join() {
   bc_->Wait();
   state_ = MigrationState::C_FINISHED;
 }
 
-void IncomingSlotMigration::StartFlow(uint32_t shard, io::Source* source) {
+void IncomingSlotMigration::Cancel() {
+  cntx_.Cancel();
+  for (auto& flow : shard_flows_) {
```
Comment (BorysTheDev) [inline src/server/cluster/incoming_slot_migration.cc:113]: flows are created in different threads, I think you need AwaitFiberOnAll(cencel_cb);

File: src/server/cluster/cluster_config.h [hunk 2]
```diff
@@ -19,6 +19,7 @@ enum class MigrationState : uint8_t {
   C_CONNECTING,
   C_SYNC,
   C_FINISHED,
+  C_CANCELLED,
```
Comment (BorysTheDev) [inline src/server/cluster/cluster_config.h:22]: why do you need it? If we cancel migration we should remove it at all

File: src/server/cluster/cluster_family.cc [hunk 3]
```diff
@@ -736,8 +738,18 @@ void ClusterFamily::RemoveOutgoingMigrations(const std::vector<MigrationInfo>& m
     auto it = std::find_if(outgoing_migration_jobs_.begin(), outgoing_migration_jobs_.end(),
                            [&m](const auto& om) { return m == om->GetMigrationInfo(); });
     DCHECK(it != outgoing_migration_jobs_.end());
+    DCHECK(it->get() != nullptr);
+    OutgoingMigration& migration = *it->get();
+    if (migration.GetState() != MigrationState::C_FINISHED) {
```
Comment (BorysTheDev) [inline src/server/cluster/cluster_family.cc:743]: we can cancel even finished migration

File: src/server/cluster/cluster_family.cc [hunk 4]
```diff
@@ -748,6 +760,30 @@ void ClusterFamily::RemoveIncomingMigrations(const std::vector<MigrationInfo>& m
...
     DCHECK(it != incoming_migrations_jobs_.end());
+    DCHECK(it->get() != nullptr);
+    IncomingSlotMigration& migration = *it->get();
+
+    // Flush non-owned migrations
+    SlotSet migration_slots(migration.GetSlots());
+    SlotSet removed = migration_slots.GetRemovedSlots(tl_cluster_config->GetOwnedSlots());
+
+    // First cancel socket, then flush slots, so that new entries won't arrive after we flush.
+    if (migration.GetState() != MigrationState::C_FINISHED) {
```
Comment (BorysTheDev) [inline src/server/cluster/cluster_family.cc:771]: state can be changed in the different thread so we don't know is it finished or not when you do cancel

File: src/server/cluster/cluster_family.cc (hunk 3 above)
Comment (BorysTheDev) [reply src/server/cluster/cluster_family.cc:743]: state can be changed in the different thread

File: src/server/cluster/incoming_slot_migration.cc [hunk 5]
```diff
@@ -85,15 +95,28 @@ IncomingSlotMigration::IncomingSlotMigration(string source_id, Service* se, Slot
...
+  LOG(INFO) << "Cancelling incoming migration of slots " << SlotRange::ToString(slots_);
+  cntx_.Cancel();
+
+  auto cb = [this](util::ProactorBase* pb) {
+    if (const auto* shard = EngineShard::tlocal(); shard) {
+      if (auto& flow = shard_flows_[shard->shard_id()]; flow) {
+        flow->Cancel();
+      }
+    }
+  };
```
Comment (BorysTheDev) [inline src/server/cluster/incoming_slot_migration.cc:115]: we can not use shard_id for incoming migrations, we don't know where every flow is executed

File: src/server/cluster/cluster_family.cc (hunk 3 above)
Comment (chakaz) [reply src/server/cluster/cluster_family.cc:743]: > we can cancel even finished migration

Not really, it requires unregistering from db slice, for which we may have not registered to (like in the case of the first config assignment)

> state can be changed in the different thread

Do you mean that there could be a race here?

File: src/server/cluster/incoming_slot_migration.cc (hunk 5 above)
Comment (chakaz) [reply src/server/cluster/incoming_slot_migration.cc:115]: you are right! good catch
furthermore, this isn't even needed, inside the inner Cancel() I forward to the correct proactor.

File: src/server/cluster/cluster_family.cc (hunk 3 above)
Comment (chakaz) [reply src/server/cluster/cluster_family.cc:743]: Both checks are done under `migration_mu_` - is there another lock I should use?

Comment (BorysTheDev) [reply src/server/cluster/cluster_family.cc:743]: migration_mu_ is used to create migrations, but state is changed inside migration

Comment (chakaz) [reply src/server/cluster/cluster_family.cc:743]: There doesn't seem to be a lock which locks the inner fiber's state changes at all, right?

File: src/server/cluster/cluster_config.h (hunk 2 above)
Comment (chakaz) [reply src/server/cluster/cluster_config.h:22]: See the usage near calling `FinishMigration()`, although I'm open to suggestions here :)

**Review by BorysTheDev — APPROVED** [review]

**Pull request merged**